### PR TITLE
 PUBDEV-3915-h2o-assembly-confusion-matrices-docs-api-tests: added do…

### PR DESCRIPTION
**Pull request by wendycwong**
PUBDEV-3915: add docs to assembly module, confusion matrices.  Add API tests for assembly module.

1.  Added documentation to assembly.py
2. Added API tests to assembly.py
3. Did not add docs to confusion matrix as it is pretty clear to me at this point.

### CODE REVIEW

File: h2o-py/h2o/assembly.py [hunk 1]
```diff
@@ -13,11 +13,57 @@
...
+    - less_than_equal
+
+    - equal_equal
+
+    - not_equal
+
+    - greater_than
+
+    - greater_than_equal
+
```
Comment (st-pasha) [inline h2o-py/h2o/assembly.py]: Please go to folder `h2o-py/docs` and run `make clean; make html`. Check that no warnings are produced, and then also look at the output `_build/index.html` in the browser. 
I have a feeling the formatting is off here...

File: h2o-py/h2o/assembly.py [hunk 2]
```diff
@@ -55,11 +101,21 @@ def names(self):
...
+        :param get_jar (bool): Whether to also download the h2o-genmodel.jar file needed to compile the POJO
+        :return: None
+        """
         assert_is_type(pojo_name, str)
         assert_is_type(path, str)
         assert_is_type(get_jar, bool)
         if pojo_name == "":
             pojo_name = "AssemblyPOJO_" + str(uuid.uuid4())
+
+        if pojo_name == "": pojo_name = "AssemblyPOJO_" + str(uuid.uuid4())
```
Comment (st-pasha) [inline h2o-py/h2o/assembly.py]: duplicate?

File: h2o-py/h2o/assembly.py [hunk 3]
```diff
@@ -55,11 +101,21 @@ def names(self):
...
         java = h2o.api("GET /99/Assembly.java/%s/%s" % (self.id, pojo_name))
         file_path = path + "/" + pojo_name + ".java"
         if path == "":
```
Comment (st-pasha) [inline h2o-py/h2o/assembly.py]: in the lines 127-131 below, replace all that nonsense with a single call to h2o.api(). Check `download_mojo` in `model_base.py` for an example

File: h2o-py/h2o/assembly.py [hunk 4]
```diff
@@ -88,9 +144,20 @@ def to_pojo(self, pojo_name="", path="", get_jar=True):
...
+        :return: H2OFrame after munging operations are completed.
+        """
         assert_is_type(fr, H2OFrame)
         steps = "[%s]" % ",".join(quoted(step[1].to_rest(step[0]).replace('"', "'")) for step in self.steps)
         j = h2o.api("POST /99/Assembly", data={"steps": steps, "frame": fr.frame_id})
+        res = []
+        for step in self.steps:
+            res.append(step[1].to_rest(step[0]))
+        res = "[" + ",".join([quoted(r.replace('"', "'")) for r in res]) + "]"
+        j = h2o.api("POST /99/Assembly", data={"steps": res, "frame": fr.frame_id})
```
Comment (st-pasha) [inline h2o-py/h2o/assembly.py]: repeated code lines 156-160

File: h2o-py/h2o/assembly.py (hunk 1 above)
Comment (wendycwong) [inline h2o-py/h2o/assembly.py]: it actually works.  I checked already.

File: h2o-py/h2o/assembly.py [hunk 5]
```diff
@@ -13,11 +13,57 @@
...
-    Extension class of Pipeline implementing additional methods:
+    H2OAssembly class can be used to specify multiple frame operations in one place.
+
+    Sample usage:
+
+    >>> my_frame = h2o.import_file(pyunit_utils.locate("smalldata/iris/iris_wheader.csv"))  # some existing H2OFrame
+    >>> assembly = H2OAssembly(steps=[("col_select",      H2OColSelect(["sepal_len", "petal_len", "class"])),
+                                       ("cos_sep_len",     H2OColOp(op=H2OFrame.cos, col="sepal_len", inplace=True)),
+                                       ("str_cnt_species", H2OColOp(op=H2OFrame.countmatches, col="class", inplace=False, pattern="s"))])
+    >>> result = assembly.fit(fr)  # fit the assembly and perform the munging operations
```
Comment (wendycwong) [inline h2o-py/h2o/assembly.py]: Done

File: h2o-py/h2o/assembly.py [hunk 6]
```diff
@@ -13,11 +13,57 @@
 
 class H2OAssembly(object):
     """
-    Extension class of Pipeline implementing additional methods:
+    H2OAssembly class can be used to specify multiple frame operations in one place.
+
+    Sample usage:
+
+    >>> my_frame = h2o.import_file(pyunit_utils.locate("smalldata/iris/iris_wheader.csv"))  # some existing H2OFrame
```
Comment (wendycwong) [inline h2o-py/h2o/assembly.py]: Done

File: h2o-py/h2o/frame.py [hunk 7]
```diff
@@ -2497,7 +2497,7 @@ def difflag1(self):
         Conduct a diff-1 transform on a numeric frame column.
 
         :returns: an H2OFrame where each element is equal to the corresponding element in the source
-        frame minus the previous-row element in the same frame.
+            frame minus the previous-row element in the same frame.
```
Comment (st-pasha) [inline h2o-py/h2o/frame.py]: (except this one)

File: h2o-py/h2o/frame.py [hunk 8]
```diff
@@ -1261,10 +1259,10 @@ def __getitem__(self, item):
...
             - a single-column boolean frame, selecting rows for which the selector is true
             - a 2-element tuple, where the first element is a row selector, and the second element is the
-              column selector. Here the row selector may be one of: an int, a list of ints, a slice, or
-              a boolean frame. The column selector is similarly one of: an int, a list of ints, a string,
-              a list of strings, or a slice. It is also possible to use the empty slice (``:``) to select
-              all elements within one of the dimensions.
+                column selector. Here the row selector may be one of: an int, a list of ints, a slice, or
+                a boolean frame. The column selector is similarly one of: an int, a list of ints, a string,
+                a list of strings, or a slice. It is also possible to use the empty slice (``:``) to select
+                all elements within one of the dimensions.
```
Comment (st-pasha) [inline h2o-py/h2o/frame.py]: it should be 2-space indent for itemlists, o/w formatting is broken.

File: h2o-py/h2o/frame.py [hunk 9]
```diff
@@ -2016,7 +2014,7 @@ def var(self, y=None, na_rm=False, use=None):
             - ``"everything"``: outputs NaNs whenever one of its contributing observations is missing
             - ``"all.obs"``: presence of missing observations will throw an error
             - ``"complete.obs"``: discards missing values along with all observations in their rows so that only
-              complete observations are used
+                complete observations are used
```
Comment (st-pasha) [inline h2o-py/h2o/frame.py]: should be 2 spaces

File: h2o-py/h2o/frame.py [hunk 10]
```diff
@@ -2057,7 +2055,7 @@ def cor(self, y=None, na_rm=False, use=None):
...
```
Comment (st-pasha) [inline h2o-py/h2o/frame.py]: should be 2 spaces

File: h2o-py/h2o/frame.py [hunk 11]
```diff
@@ -2639,15 +2637,15 @@ def stratified_split(self, test_frac=0.2, seed=-1):
...
-          >>> test["y"].table()["Count"] / test["y"].table()["Count"].sum()
+            >>> stratsplit = df["y"].stratified_split(test_frac=0.3, seed=12349453)
+            >>> train = df[stratsplit=="train"]
+            >>> test = df[stratsplit=="test"]
+            >>>
+            >>> # check that the distributions among the initial frame, and the
+            >>> # train/test frames match
+            >>> df["y"].table()["Count"] / df["y"].table()["Count"].sum()
+            >>> train["y"].table()["Count"] / train["y"].table()["Count"].sum()
+            >>> test["y"].table()["Count"] / test["y"].table()["Count"].sum()
```
Comment (st-pasha) [inline h2o-py/h2o/frame.py]: should be 2 spaces

File: h2o-py/tests/testdir_apis/H2OAssembly/pyunit_h2oassembly_divide.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
+from __future__ import print_function
+import sys
+sys.path.insert(1,"../../")
+from tests import pyunit_utils
+from h2o.assembly import *
+from h2o.utils.typechecks import assert_is_type
+
+def h2oassembly_divide():
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/H2OAssembly/pyunit_h2oassembly_divide.py]: seems like tests for 
    divide = H2OFrame.__truediv__
    plus = H2OFrame.__add__
    multiply = H2OFrame.__mul__
    minus = H2OFrame.__sub__
    less_than = H2OFrame.__lt__
    less_than_equal = H2OFrame.__le__
    equal_equal = H2OFrame.__eq__
    not_equal = H2OFrame.__ne__
    greater_than = H2OFrame.__gt__
    greater_than_equal = H2OFrame.__ge__

can all be merged into a single testcase

File: h2o-py/tests/testdir_apis/H2OAssembly/pyunit_h2oassembly_to_pojo.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+                       col_types=["numeric","numeric","numeric","numeric","string"])
+    assembly = H2OAssembly(steps=[("col_select",      H2OColSelect(["sepal_len", "petal_len", "class"])),
+                                ("cos_sep_len",     H2OColOp(op=H2OFrame.cos, col="sepal_len", inplace=True)),
+                                ("str_cnt_species",
+                                 H2OColOp(op=H2OFrame.countmatches, col="class", inplace=False, pattern="s"))])
+
+    result = assembly.fit(fr)  # fit the assembly
+    assert_is_type(result, H2OFrame)
+
+    try:
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/H2OAssembly/pyunit_h2oassembly_to_pojo.py]: avoid try-catch, instead check whether `results_dir` exists explicitly (`os.path.isdir`) and do the correct action based on that

File: h2o-py/h2o/assembly.py [hunk 14]
```diff
@@ -55,6 +100,14 @@ def names(self):
...
+        Convert the munging operations performed on H2OFrame into a POJO.
+
+        :param pojo_name (str): Name of POJO
+        :param path (str): path of POJO.
+        :param get_jar (bool): Whether to also download the h2o-genmodel.jar file needed to compile the POJO
+        :return: None
+        """
         assert_is_type(pojo_name, str)
         assert_is_type(path, str)
         assert_is_type(get_jar, bool)
```
Comment (wendycwong) [inline h2o-py/h2o/assembly.py]: Good one!

File: h2o-py/tests/testdir_apis/H2OAssembly/pyunit_h2oassembly_divide.py (hunk 12 above)
Comment (wendycwong) [inline h2o-py/tests/testdir_apis/H2OAssembly/pyunit_h2oassembly_divide.py]: Done.

File: h2o-py/tests/testdir_apis/H2OAssembly/pyunit_h2oassembly_to_pojo.py (hunk 13 above)
Comment (wendycwong) [inline h2o-py/tests/testdir_apis/H2OAssembly/pyunit_h2oassembly_to_pojo.py]: Done.

File: h2o-py/h2o/assembly.py [hunk 15]
```diff
@@ -67,13 +110,8 @@ def to_pojo(self, pojo_name="", path="", get_jar=True):
...
             with open(file_path, 'w', encoding="utf-8") as f:
                 f.write(java)  # this had better be utf-8 ?
-        if get_jar and path != "":
-            url = h2o.connection().make_url("h2o-genmodel.jar")
-            filename = path + "/" + "h2o-genmodel.jar"
-            response = urlopen()(url)
-            with open(filename, "wb") as f:
-                f.write(response.read())
-
+        if get_jar and path != "":      # incorporate Pasha code review suggestion here to not use urlopen...
```
Comment (st-pasha) [inline h2o-py/h2o/assembly.py]: remove comment

**Comment by wendycwong** [comment]
All pass!

**Pull request merged**